### PR TITLE
Reduce string allocations when formatting file system objects.

### PR DESCRIPTION
**Pull request by powercode**
## PR Summary

Cache FullName property in the ProviderInfo class. Causes extra string allocations for every item in the formatting pipeline.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.  
- **User-facing changes**
    - [x] Not Applicable
    - **OR**  
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
        - [ ] [Add `[feature]` to your commit messages if the change is significant or affects feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)

### CODE REVIEW

**Review by adityapatwardhan — APPROVED** [review]

**Comment by adityapatwardhan** [comment]
@powercode Please have a look at the test failure.

**Review by iSazonov — APPROVED** [review]

**Comment by iSazonov** [comment]
@powercode Could you please fix CI Windows?

File: src/System.Management.Automation/engine/DataStoreAdapterProvider.cs [hunk 1]
```diff
@@ -136,6 +143,7 @@ public string ModuleName
         internal void SetModule(PSModuleInfo module)
         {
             Module = module;
+            _fullName = null; // clear the cached FullName
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/DataStoreAdapterProvider.cs]: Please remove the comment. Better use `_cacheFullModuleName` name.

Comment (powercode) [reply src/System.Management.Automation/engine/DataStoreAdapterProvider.cs]: But it isn't a full module name. That name would be misleading. But sure, I can remove the comment.

Comment (iSazonov) [reply src/System.Management.Automation/engine/DataStoreAdapterProvider.cs]: If it is not a full name what about _cachedModuleName?

**Comment by bergmeister** [comment]
Awesome. Out pf curiosity: Do you have time measurements  of cases where this improvement is noticeable?

**Comment by powercode** [comment]
@bergmeister It is mostly in allocations - it generates lots of extra strings, like 2 strings per item in the pipeline.

I'm reasoning more in terms of steadily removing silly allocations, and fixing the worst offenders perf-wise, and eventually we will have a more snappy shell.

The image below show the dotmemory view of strings allocated, grouped by callstack, when looking at the first 50000 items in the windows dir. It wasn't the biggest, but it was one of the bigger, and it was low hanging fruit. Does it make sense?

image

File: src/System.Management.Automation/engine/DataStoreAdapterProvider.cs (hunk 1 above)
Comment (powercode) [reply src/System.Management.Automation/engine/DataStoreAdapterProvider.cs]: The property is FullName, so it makes sense to me that the backing property is _fullName.

**Comment by bergmeister** [comment]
Thanks @powercode. Yes, makes sense. I agree that continous improvement will make it better over time. :)

**Comment by powercode** [comment]
@bergmeister It may seem like I'm just doing random changes, but I actually measure things now and then :)

File: src/System.Management.Automation/engine/DataStoreAdapterProvider.cs (hunk 1 above)
Comment (iSazonov) [reply src/System.Management.Automation/engine/DataStoreAdapterProvider.cs]: I tried to insert your comment to the name that it is a cached :-)

**Comment by bergmeister** [comment]
@powercode I was not questioning it, I was just curious for my own education (because knowing this also means that one will know the scenarios where Windows PowerShell is weak and where I could recommend an upgrade to clients).
Would you mind having a look at the following issue with Import-Csv please which is causing actual OutOfMemory problems in some workflows where CSV files are a couple of GB large.
https://github.com/PowerShell/PowerShell/issues/7603

**Comment by iSazonov** [comment]
@bergmeister Windows PowerShell is still more efficient than PowerShell Core 
in many scenarios. Main reason that .Net Framework engine works differently than .Net Core one. Optimizing individual cmdlets would be fine, but maybe we need something more general.

**Comment by powercode** [comment]
@bergmeister I'm currently working on optimizations for the formatting system, the filesystem provider, etc that makes it out-perform windows powershell with a huge margin, often like 4x. And with a memory footprint reduction of the similar size.

Import-Csv is problematic, since our property abstraction is heavier that I would like it to be. Especially in the case like that, where we actually have a table, we wouldn't need to store all the metadata for each object. I haven't given it so much thought - just passed by it last fall when doing related work.

I have an idea that I like to try out for Import-CSV - could drastically reduce the memory footprint.
That is to generate a dynamic assembly, with a class containing the fields of the CSV.

**Comment by iSazonov** [comment]
I was thinking about compiling to classes too and found some problems in the approach. One problem could be solved by #8852 (not trivial). After finding IDataView  #8855 I suppose this is preferred way we should start research.

**Pull request merged**